File: py_ferreus_rbf/examples/isosurface_linear_rmt.py
from pathlib import Path

import numpy as np
import pandas as pd
from ferreus_rbf.interpolant_config import (
    FittingAccuracy,
    FittingAccuracyType,
    InterpolantSettings,
    RBFKernelType,
)
from ferreus_rbf.isosurfacing import build_isosurface
from ferreus_rbf.progress import (
    DuplicatesRemoved,
    Message,
    Progress,
    ProgressEvent,
    SolverIteration,
    SurfacingProgress,
)

from ferreus_rbf import RBFInterpolator


def on_progress(event: ProgressEvent) -> None:
    if isinstance(event, SolverIteration):
        print(
            f"Iteration: {event.iter:3d}  {event.residual:>.5E}  {(event.progress * 100):.1f}%"
        )
    elif isinstance(event, SurfacingProgress):
        print(
            f"Isovalue: {event.isovalue}  Stage: {event.stage}  {(event.progress * 100):.1f}%"
        )
    elif isinstance(event, DuplicatesRemoved):
        print(f"Removed duplicates: {event.num_duplicates}")
    elif isinstance(event, Message):
        print(event.message)


# Define the input path for the example signed distance points
project_root = Path(__file__).resolve().parents[2]
pointset_path = project_root / "datasets" / "albatite_SD_points.csv"

# Import the example points file
df = pd.read_csv(pointset_path)

# Extract the source point coordinates and signed distance values
#
# The dataset contains 3D point coordinates in the first 3 columns, followed by
# signed distance (SD) values in the 4th column. The SD values are 0 on the surface
# boundary, -ve inside the surface and +ve outside the surface.
source_points = df[["X", "Y", "Z"]].to_numpy().astype(np.float64)
source_values = df["SignedDistance"].to_numpy().astype(np.float64)

# Get the axis aligned bounding box extents of the source points
# to use for the isosurface extraction
extents = np.concatenate(
    (
        np.floor(np.min(source_points, axis=0)),
        np.ceil(np.max(source_points, axis=0)),
    )
)

# Define the RBF kernel to use
kernel_type = RBFKernelType.Linear

# Define the desired fitting accuracy
fitting_accuracy = FittingAccuracy(0.01, FittingAccuracyType.Absolute)

# Initialise an InterpolantSettings instance
interpolant_settings = InterpolantSettings(
    kernel_type, fitting_accuracy=fitting_accuracy
)

# Create a callback to receive progress updates from the RBFInterpolator
callback = Progress(callback=on_progress)

# Setup and solve the RBF system
rbfi = RBFInterpolator(
    source_points, source_values, interpolant_settings, progress_callback=callback
)

# Define the sampling grid resolution for the surfacer
resolution = 5

# When setting up the rbf evaluator for isosurfacing we need to add a buffer to the
# extents of the evaluator so we don't end up trying to evaluate points outside the
# extents of the evaluator.
evaluator_extents = np.hstack(
    (
        extents[:3] - resolution * 10.0,
        extents[3:] + resolution * 10.0,
    )
)

# Build the RBF evaluator
rbfi.build_evaluator(evaluator_extents)

# Define the isosurfacing evaluation function to use the rbf evaluator
isosurface_fn = lambda targets: rbfi.evaluate_targets(targets)

# Define the isovalue at which to surface
isovalue = 0.0

# Extract the isosurface
mesh = build_isosurface(
    source_points,
    extents,
    resolution,
    isovalue,
    isosurface_fn,
)

# Save the isosurface out to an obj file
surface_name = f"isosurface_linear_{resolution}m"
outpath = Path(f"{surface_name}.obj")
mesh.save_obj(str(outpath), surface_name)

File: ferreus_rbf/examples/isosurface_linear_rmt.rs
/////////////////////////////////////////////////////////////////////////////////////////////
//
// Example 3D isosurface generation with 35,801 input signed distance points using the linear
// RBF kernel and the default constant drift.
//
// Created on: 15 Nov 2025     Author: Daniel Owen
//
// Copyright (c) 2025, Maptek Pty Ltd. All rights reserved. Licensed under the MIT License.
//
/////////////////////////////////////////////////////////////////////////////////////////////

use faer::{Mat, MatRef};
use ferreus_rbf::{
    RBFInterpolator, csv_to_point_arrays,
    interpolant_config::{
        FittingAccuracy, FittingAccuracyType, InterpolantSettings, RBFKernelType,
    },
    isosurfacing::{BoundaryClosure, ClusterMethod, build_isosurface},
    progress::{ProgressMsg, ProgressSink, ProgressSinkExt, closure_sink},
};
use std::{cell::RefCell, env, path::Path, rc::Rc, sync::Arc};

/// Nice float formatter for filenames: trims trailing zeros and dots.
fn fmt_num(x: f64) -> String {
    let s = format!("{:.6}", x);
    s.trim_end_matches('0').trim_end_matches('.').to_string()
}

/// Generates a callback closure_sink
fn get_callback_sink() -> Arc<dyn ProgressSink> {
    let (sink, _listener) = closure_sink(256, |msg| match msg {
        ProgressMsg::SolverIteration {
            iter,
            residual,
            progress,
        } => {
            println!(
                "Iteration: {:>3}    {:>.5E}    {:>.1}%",
                iter,
                residual,
                progress * 100.0
            );
        }
        ProgressMsg::SurfacingProgress {
            isovalue,
            stage,
            progress,
        } => {
            println!(
                "Isovalue: {:?}    Stage: {}    {:>.1}%",
                isovalue,
                stage,
                progress * 100.0
            );
        }
        ProgressMsg::DuplicatesRemoved { num_duplicates } => {
            println!("Removed {:>3} duplicate points", num_duplicates);
        }

        ProgressMsg::Message { message } => {
            println!("{message}");
        }
    });

    sink
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let cwd = env::current_dir().unwrap();

    // Define the filepath for the albatite test dataset
    let file_path = Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("workspace root")
        .join("datasets")
        .join("albatite_SD_points.csv");

    // Extract the source point coordinates and signed distance values
    let (source_points, source_values): (Mat<f64>, Mat<f64>) =
        csv_to_point_arrays(file_path.to_str().expect("valid UTF-8 path"), true)?;

    // Get the axis aligned bounding box extents of the source points
    // to use for the isosurface extraction
    let source_point_extents = ferreus_rbf_utils::get_pointarray_extents(source_points.as_ref());

    // Define the RBF kernel to use
    let kernel_type = RBFKernelType::Linear;

    // Define the desired fitting accuracy
    let fitting_accuracy = FittingAccuracy {
        tolerance: 0.01,
        tolerance_type: FittingAccuracyType::Absolute,
    };

    // Initialise an InterpolantSettings instance
    let interpolant_settings = InterpolantSettings::builder(kernel_type)
        .fitting_accuracy(fitting_accuracy)
        .build();

    // Create a callback to receive progress updates from the RBFInterpolator
    let callback = get_callback_sink();

    // Setup and solve the RBF system
    let mut rbfi = RBFInterpolator::builder(
        source_points.clone(),
        source_values.clone(),
        interpolant_settings,
    )
    .progress_callback(callback.clone())
    .build();

    // Define the sampling grid resolution for the surfacer
    let resolution = 10.0;

    let evaluator_extents: Vec<f64> = source_point_extents
        .iter()
        .enumerate()
        .map(|(idx, val)| match idx < 3 {
            true => val - resolution * 10.0,
            false => val + resolution * 10.0,
        })
        .collect();

    rbfi.build_evaluator(Some(evaluator_extents));

    let rbfi = Rc::new(RefCell::new(rbfi));
    let rbfi_surface = Rc::clone(&rbfi);
    let rbfi_grad = Rc::clone(&rbfi);

    let mut surface_fn =
        move |targets: MatRef<f64>| rbfi_surface.borrow_mut().evaluate_targets(targets.as_ref());

    let mut gradient_fn = move |targets: MatRef<f64>| {
        rbfi_grad
            .borrow_mut()
            .evaluate_targets_with_gradients(targets.as_ref())
    };

    let rmt_callback = callback.clone().into_rmt_progress();

    //  Define the isovalue at which to surface
    let isovalue = 0.0;

    // Generate an isosurface directly from the RMT extractor.
    let mesh = build_isosurface(
        source_points.as_ref(),
        &source_point_extents,
        resolution,
        isovalue,
        &mut surface_fn,
        Some(&mut gradient_fn),
        ClusterMethod::CurvatureWeighted,
        BoundaryClosure::ClosePositive,
        Some(rmt_callback.as_ref()),
    );

    //Save the isosurface out to an obj file
    let name = format!("isosurface_linear_{}m", fmt_num(resolution));
    let outpath = cwd.join("examples").join(format!("{}.obj", name));
    mesh.save_obj(outpath, &name)?;

    Ok(())
}
